### PR TITLE
Update the scripts used to generate the ucsc-XYZ recipes

### PR DESCRIPTION
**Pull request by martin-g**
This PR updates the scripts used to generate the `ucsc-xyz` recipes. 
See https://github.com/bioconda/bioconda-recipes/pull/50323 and https://github.com/bioconda/bioconda-recipes/pull/50324/ for examples of the changes in the recipes.

Once this PR is merged the other recipes may be pushed in bulk ?!

----

Please read the [guidelines for Bioconda recipes](https://bioconda.github.io/contributor/guidelines.html) before opening a pull request (PR).

### General instructions

* If this PR adds or updates a recipe, use "Add" or "Update" appropriately as the first word in its title.
* New recipes not directly relevant to the biological sciences need to be submitted to the [conda-forge channel](https://conda-forge.org/docs/) instead of Bioconda.
* PRs require reviews prior to being merged. Once your PR is passing tests and ready to be merged, please issue the `@BiocondaBot please add label` command.
* Please post questions [on Gitter](https://gitter.im/bioconda/Lobby) or ping `@bioconda/core` in a comment.

### Instructions for avoiding API, ABI, and CLI breakage issues
Conda is able to record and lock (a.k.a. pin) dependency versions used at build time of other recipes.
This way, one can avoid that expectations of a downstream recipe with regards to API, ABI, or CLI are violated by later changes in the recipe.
If not already present in the meta.yaml, make sure to specify `run_exports` (see [here](https://bioconda.github.io/contributor/linting.html#missing-run-exports) for the rationale and comprehensive explanation).
Add a `run_exports` section like this:

```yaml
build:
  run_exports:
    - ...

```

with `...` being one of:

| Case                             | run_exports statement                                               |
| -------------------------------- | ------------------------------------------------------------------- |
| semantic versioning              | `{{ pin_subpackage("myrecipe", max_pin="x") }}`     |
| semantic versioning (0.x.x)      | `{{ pin_subpackage("myrecipe", max_pin="x.x") }}`   |
| known breakage in minor versions | `{{ pin_subpackage("myrecipe", max_pin="x.x") }}` (in such a case, please add a note that shortly mentions your evidence for that) |
| known breakage in patch versions | `{{ pin_subpackage("myrecipe", max_pin="x.x.x") }}` (in such a case, please add a note that shortly mentions your evidence for that) |
| calendar versioning              | `{{ pin_subpackage("myrecipe", max_pin=None) }}`    |

while replacing `"myrecipe"` with either `name` if a `name|lower` variable is defined in your recipe or with the lowercase name of the package in quotes.

### Bot commands for PR management

<details>
  <summary>Please use the following BiocondaBot commands:</summary>

Everyone has access to the following BiocondaBot commands, which can be given in a comment:

<table>
  <tr>
    <td><code>@BiocondaBot please update</code></td>
    <td>Merge the master branch into a PR.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please add label</code></td>
    <td>Add the <code>please review & merge</code> label.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please fetch artifacts</code></td>
    <td>Post links to CI-built packages/containers. <br />You can use this to test packages locally.</td>
  </tr>
</table>

Note that the <code>@BiocondaBot please merge</code> command is now depreciated. Please just squash and merge instead.

Also, the bot watches for comments from non-members that include `@bioconda/<team>` and will automatically re-post them to notify the addressed `<team>`.

</details>

### CODE REVIEW

**Comment by martin-g** [comment]
@bgruening This is the result of your suggestion at https://github.com/bioconda/bioconda-recipes/pull/50106#issuecomment-2298671756

What would be the best way to rebuild the recipes ? 
- should I open PRs against `master` in mini-bulks (i.e. 5-6 recipes in one PR) ?
- should someone (@aliciaaevans ?!) do it in the `bulk` branch ?

File: scripts/ucsc/template-meta-with-python.yaml [hunk 1]
```diff
@@ -23,14 +26,14 @@ requirements:
     - libpng
     - libuuid
     - mysql-connector-c
-    - openssl
+    - libopenssl-static
```
Comment (bgruening) [inline scripts/ucsc/template-meta-with-python.yaml:29]: interesting why is this needed?

File: scripts/ucsc/template-meta-with-python.yaml [hunk 2]
```diff
@@ -43,3 +46,9 @@ about:
   home: "http://hgdownload.cse.ucsc.edu/admin/exe/"
   license: "varies; see http://genome.ucsc.edu/license"
   summary: "{summary}"
+  dev_url: "https://github.com/ucscGenomeBrowser/kent"
```
Comment (bgruening) [inline scripts/ucsc/template-meta-with-python.yaml:49]: cool thanks!

**Comment by aliciaaevans** [comment]
Oh cool, I didn't realize these scripts existed. I could run this on bulk once it's merged. Should be relatively quick since these aren't interdependent like bioconductor recipes.

File: scripts/ucsc/template-meta-with-python.yaml [hunk 3]
```diff
@@ -43,3 +46,9 @@ about:
...
+  doc_url: "https://github.com/ucscGenomeBrowser/kent/blob/master/README"
+
+extra:
+  additional-platforms:
```
Comment (bgruening) [inline scripts/ucsc/template-meta-with-python.yaml:53]: Should we add a doi as reference here: https://genome.ucsc.edu/cite.html#ref

Maybe the bigwig, bigbed paper? https://doi.org/10.1093/bioinformatics/btq351

File: scripts/ucsc/template-build-cp.sh [hunk 4]
```diff
@@ -1,4 +1,7 @@
 #!/bin/bash
+
+set -xe
+
 mkdir -p "$PREFIX/bin"
-cp {program_source_dir}/{program} "$PREFIX/bin"
-chmod +x "$PREFIX/bin/{program}"
+cp -f {program_source_dir}/{program} "${{PREFIX}}/bin"
```
Comment (bgruening) [inline scripts/ucsc/template-build-cp.sh:6]: Is the force needed, or are we masking potential error here?

**Review by bgruening — COMMENTED** [review]
Awesome, thanks a lot!

**Comment by bgruening** [comment]
This is super cool, thanks for updating them. We should use the `bulk` branch here, that why we have it. Mini-bulks are just wasting your time.

**Comment by martin-g** [comment]
Thanks, @aliciaaevans !

There are ~200 disabled ucsc- recipes in `build-fail-blacklist` that should be re-enabled before running the bulk build!
There will be (hopefully few) recipes like https://github.com/bioconda/bioconda-recipes/pull/49297 which will fail on `linux-aarch64` - for them we will need to remove the `additional-platforms` entry. Actually I could add a new array in the Python script for the known ones, so they are skipped automatically. Let me update the PR!

File: scripts/ucsc/template-build-cp.sh (hunk 4 above)
Comment (martin-g) [reply scripts/ucsc/template-build-cp.sh:6]: This is an improvement by @mencian to one of my previous PRs.
I could remove it if there is a chance to mask some problem!

File: scripts/ucsc/template-meta-with-python.yaml (hunk 1 above)
Comment (martin-g) [reply scripts/ucsc/template-meta-with-python.yaml:29]: The tools require libssl.a instead of libssl.so. I am not sure since when though.
I have seen some recipes using `includes.patch` to change `.a` to `.so`. I decided to keep the original build, i.e. to not patch it, and use the new dependency instead.

File: scripts/ucsc/template-meta-with-python.yaml (hunk 3 above)
Comment (martin-g) [reply scripts/ucsc/template-meta-with-python.yaml]: Could you please give me the YAML snippet that I should add? Or please push it yourself to this PR!

**Comment by martin-g** [comment]
> Actually I could add a new array in the Python script for the known ones, so they are skipped automatically. Let me update the PR!

Done!

File: scripts/ucsc/template-build-cp.sh (hunk 4 above)
Comment (bgruening) [reply scripts/ucsc/template-build-cp.sh:6]: I'm not sure about the reason, so I would rather remove.

File: scripts/ucsc/template-meta.yaml [hunk 5]
```diff
@@ -8,37 +8,47 @@ package:
...
-  home: "http://hgdownload.cse.ucsc.edu/admin/exe/"
-  license: "varies; see http://genome.ucsc.edu/license"
+  home: "https://hgdownload.cse.ucsc.edu/admin/exe/"
+  license: "Varies; see http://genome.ucsc.edu/license"
   summary: "{summary}"
+  dev_url: "https://github.com/ucscGenomeBrowser/kent"
+  doc_url: "https://github.com/ucscGenomeBrowser/kent/blob/master/README"
+
+extra:
+  {linux_aarch64}
```
Comment (bgruening) [inline scripts/ucsc/template-meta.yaml:53]:
```suggestion
  {linux_aarch64}
  identifiers:
    - biotools:UCSC_Genome_Browser_Utilities
    - doi:10.1093/bioinformatics/btq351
```

File: scripts/ucsc/template-meta-with-python.yaml [hunk 6]
```diff
@@ -43,3 +46,12 @@ about:
   home: "http://hgdownload.cse.ucsc.edu/admin/exe/"
   license: "varies; see http://genome.ucsc.edu/license"
   summary: "{summary}"
+  dev_url: "https://github.com/ucscGenomeBrowser/kent"
+  doc_url: "https://github.com/ucscGenomeBrowser/kent/blob/master/README"
+
+extra:
+  identifiers:
+    - biotools:UCSC_Genome_Browser_Utilities
+    - doi:10.1093/bioinformatics/btq351
```
Comment (bgruening) [inline scripts/ucsc/template-meta-with-python.yaml:55]:
```suggestion
    - doi:10.1093/bib/bbs038
```

File: scripts/ucsc/template-meta.yaml [hunk 7]
```diff
@@ -8,37 +8,50 @@ package:
...
+  home: "https://hgdownload.cse.ucsc.edu/admin/exe/"
+  license: "Varies; see http://genome.ucsc.edu/license"
   summary: "{summary}"
+  dev_url: "https://github.com/ucscGenomeBrowser/kent"
+  doc_url: "https://github.com/ucscGenomeBrowser/kent/blob/master/README"
+
+extra:
+  identifiers:
+    - biotools:UCSC_Genome_Browser_Utilities
+    - doi:10.1093/bioinformatics/btq351
```
Comment (bgruening) [inline scripts/ucsc/template-meta.yaml:55]:
```suggestion
    - doi:10.1093/bib/bbs038
```

**Review by bgruening — COMMENTED** [review]
Found a better one, sorry.

**Comment by martin-g** [comment]
@BiocondaBot please add label

File: scripts/ucsc/create-ucsc-packages.py [hunk 8]
```diff
@@ -198,6 +197,12 @@ def program_subdir(program, names):
     'gfServer',
 ]
 
+# A list of programs which have problems to build on linux-aarch64
+SKIP_AARCH64 = [
```
Comment (bgruening) [inline scripts/ucsc/create-ucsc-packages.py:201]: I think at some point we should have a global list, that script like this one can consume.

**Review by bgruening — APPROVED** [review]
Thanks a lot @martin-g

**Comment by aliciaaevans** [comment]
@martin-g I've got time today to get this started on bulk.

**Comment by aliciaaevans** [comment]
Still working though timeouts and disk space limits, but here are failures so far: https://github.com/bioconda/bioconda-recipes/wiki/build-failures-bulk 

I think the aarch failures are all that `knet_init_alt` so we can remove the platform for those after they've all been built or failed. I think the linux-64 are mostly one error: `multiple definition of htmlRecover`.

**Comment by aliciaaevans** [comment]
(I'm still using this PR for notes about this bulk update for ucsc.)

I've removed `linux-aarch64` from the failures with the `knet_init_alt` error, removed the build_failure files, and added them to the `SKIP_AARCH64` in the python script (all on the `bulk` branch). Now https://github.com/bioconda/bioconda-recipes/wiki/build-failures-bulk only has other failures. I will probably merge bulk back to master sooner rather than later (tomorrow?). We can always do another bulk run to fix the `knet_init_alt` issue and the remaining failures can be individual PRs.

**Comment by martin-g** [comment]
Thank you, @aliciaaevans !

In https://github.com/bioconda/bioconda-recipes/pull/50368 @mencian added support for osx-64 to one of the ucsc recipes.
After running the scripts in bulk this change will be overwritten and lost.
@mencian We can add the OSX specific lines to the templates once @aliciaaevans merges the current improvements back to `master`.
